sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,635 @@
+
+    public Write withBatchSize(long batchSize) {
+      return builder().setBatchSize(batchSize).build();
+    }
+
+    public Write withBatchSizeMegaBytes(int batchSizeMegaBytes) {
+      return builder().setBatchSizeMegaBytes(batchSizeMegaBytes).build();
+    }
+
+    @Override
+    public void validate(PCollection<String> input) {
+      checkState(getConnectionConfiguration() != null, "ElasticsearchIO.write() requires a "
+          + "connection configuration to be set via withConnectionConfiguration(configuration)");
+    }
+
+    @Override
+    public PDone apply(PCollection<String> input) {
+      input.apply(ParDo.of(new WriterFn(this)));
+      return PDone.in(input.getPipeline());
+    }
+
+    private static class WriterFn extends DoFn<String, Void> {
+
+      private final Write spec;
+
+      private transient RestClient restClient;
+      private ArrayList<String> batch;
+      private long currentBatchSizeBytes;
+
+      public WriterFn(Write spec) {
+        this.spec = spec;
+      }
+
+      @Setup
+      public void createClient() throws Exception {
+        restClient = spec.getConnectionConfiguration().createClient();
+      }
+
+      @StartBundle
+      public void startBundle(Context context) throws Exception {
+        batch = new ArrayList<>();
+        currentBatchSizeBytes = 0;
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext context) throws Exception {
+        String document = context.element();
+        batch.add(String.format("{ \"index\" : {} }%n%s%n", document));
+        currentBatchSizeBytes += document.getBytes().length;
+        if (batch.size() >= spec.getBatchSize()
+            || currentBatchSizeBytes >= (spec.getBatchSizeMegaBytes() * 1024L * 1024L)) {
+          finishBundle(context);
+        }
+      }
+
+      @FinishBundle
+      public void finishBundle(Context context) throws Exception {
+        StringBuilder bulkRequest = new StringBuilder();
+        if (batch.size() > 0) {
+          for (String json : batch) {
+            bulkRequest.append(json);
+          }
+          Response response;
+          String endPoint = String.format("/%s/%s/_bulk",
+                                          spec.getConnectionConfiguration().getIndex(),
+                                          spec.getConnectionConfiguration().getType());
+          HttpEntity requestBody = new NStringEntity(bulkRequest.toString(), ContentType
+              .APPLICATION_JSON);
+          response =
+              restClient.performRequest("POST", endPoint, Collections.<String, String>emptyMap(),
+                                        requestBody, new
+                                            BasicHeader("", ""));
+          JsonObject searchResult = parseResponse(response);
+          boolean errors = searchResult.getAsJsonPrimitive("errors").getAsBoolean();
+          if (errors) {
+              StringBuilder errorMessages = new StringBuilder("Error writing to Elasticsearch, "
```

Comment:
Indentation here is off, too. And in a bunch more places. And a bunch of whitespace is also off (e.g. line 607).

I usually run my code through [google-java-format](https://github.com/google/google-java-format) to stop worrying about these issues. It's not specifically a mandated code style, but at least it gives a reasonable formatting, consistency and convenience (it has IDE plugins too).